design-documents/csp.md

```diff
@@ -0,0 +1,208 @@
+to the server but proceed as it would without these restrictions. This mode would be convenient to merchants
+just adopting CSP on their existing stores to see what kind of legit resources/features they need to enable
+via CSP.
+## Backward compatibility
+In order to remind merchants that having CSP defined is important it will be enabled by default. But since adopting it
+would require manually whitelisting trusted resource/features by merchants and extension developers CSP
+will be functioning in the report-only mode. This will allow us to introduce CSP in a patch release.
+ 
+## API
+At the beginning Magento would not manage all of the possible CSP policies, that's
+why we need an extensible SPI to allow 3rd party developers to implement and control new
+policies.
+ 
+CSP API/SPI will consist of the following interfaces:
+ 
+* Policy DTO. The interface has only policy ID because not all CSP policies look like _\<policy name\> \<value\> \<value\>_,
+  some may consist from multiple headers and have different syntax
+```php
+namespace Magento\Security\Model\CSP\Data;
+
+interface PolicyInterface
+{
+    //Policy ID, like 'default-src' or 'referrer'
+    public function getId(): string;
+}
+```
+ 
+* Policy collector, will collect policies from different sources, will be composed from multiple collectors.
+```php
+namespace Magento\Security\Model\CSP;
+
+interface PolicyCollectorInterface
+{
+    //Keys - policy IDs.
+    public function collect(PolicyInterface[] $defaultPolicies = []): PolicyInterface[];
+}
+```
+ 
+* Render given policy and apply to given response object.
+```php
+namespace Magento\Security\Model\CSP;
+
+interface PolicyRendererInterface
+{
+    public function render(PolicyInterface $policy, \Magento\Framework\App\Response\HttpInterface $response): void
+}
+```
+ 
+* A violation recorded by a browser
+```php
+namespace Magento\Security\Model\CSP\Data;
+
+interface ViolationInterface
+{
+    public function getDocumentUri(): string;
+    
+    public function getViolatedDirective(): string;
+    
+    public function getOriginalPolicy(): string;
+    
+    public function getReferrer(): ?string;
+    
+    public function getBlockedUri(): ?string;
+    
+    public function getSourceJsFileLine(): ?string;
+    
+    public function getReported(): \DateTime
+}
+```
+ 
+* A repository of violations logs collected when CSP reporting is enabled
+```php
+namespace Magento\Security\Model\CSP;
+
+interface ViolationsRepositoryInterface
+{
+    public function getList(SearchCriteriaInterface $search): ViolationSearchResultInterface;
+    
+    public function create(string $reportJson): void;
+}
+```
+ 
+## CSP configuration
+Security settings like CSP should be only manageable with physical files and
+CLI requiring highest level of access.
+ 
+CSP mode will be set with a config path not appearing in the admin configuration page so merchants
+would have to execute a CLI command in order to change it.
+
```

Comment:
This sounds great: 
It would protect the settings against changes via Admin (admins, hacked accounts, admin laptops with malware, ...). Only Dev/DevOps level could change it.
Is there a way to protect the configuration against badly written extensions, too?

Reply:
Any possible problems with that on Magento Cloud? Related to read-only file system.

Reply:
It would need to be written to the env.php file during the build process before deployment completes.

Reply:
the hash comment is very good. can it be pre-generated once during deployment or the inline scripts are dynamic - if the latter it might cause caching issues too? definitely seems like an easier approach than nonces.

Reply:
@piotrekkaminski I've added a bit about hashes